jsonclient/client.go

```diff
@@ -41,6 +44,69 @@ type JSONClient struct {
 	httpClient *http.Client          // used to interact with the server via HTTP
 	Verifier   *ct.SignatureVerifier // nil for no verification (e.g. no public key available)
 	logger     Logger                // interface to use for logging warnings and errors
+
+	multiplier int64
+	until      *time.Time
+	mu         sync.RWMutex
+}
+
+const (
+	maxBackoff = time.Second * 128
+	jitter     = 250
+)
+
+func (c *JSONClient) setBackoff(override *time.Duration) time.Duration {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if c.until != nil {
+		return time.Until(*c.until)
+	}
+	var wait time.Duration
+	if override != nil {
+		wait = *override
+	} else {
+		c.multiplier++
+		wait = time.Second * time.Duration(math.Pow(2, float64(c.multiplier-1)))
+		if wait > maxBackoff {
+			wait = maxBackoff
+		}
+	}
+	until := time.Now().Add(wait)
+	c.until = &until
+	go func() {
+		time.Sleep(wait)
+		c.mu.Lock()
+		defer c.mu.Unlock()
+		c.until = nil
```

Comment:
Hmm, this relies on nothing else having changed `c.until` in the intervening interval, which I don't think is safe -- at the point when this goroutine gets kicked off, say with a 2s wailt, another goroutine might be blocked at the top of a `setBackoff(60s)` which will change `c.until` to 60s ... but `c.until` will get cleared 2s later.

Might be safer to just have `backoff()` cope with `notBefore` / `until` being in the past.